base/file.jl

```diff
@@ -88,19 +88,30 @@ end
 mkdir(path::AbstractString, mode::Signed) = throw(ArgumentError("mode must be an unsigned integer; try 0o$mode"))
 mkpath(path::AbstractString, mode::Signed) = throw(ArgumentError("mode must be an unsigned integer; try 0o$mode"))
 
-function rm(path::AbstractString; recursive::Bool=false)
+function rm(path::AbstractString; force::Bool=false, recursive::Bool=false)
     if islink(path) || !isdir(path)
-        @windows_only if (filemode(path) & 0o222) == 0; chmod(path, 0o777); end # is writable on windows actually means "is deletable"
-        unlink(path)
+        try
+            @windows_only if (filemode(path) & 0o222) == 0; chmod(path, 0o777); end # is writable on windows actually means "is deletable"
+            unlink(path)
+        catch err
+            if isa(err, UVError)
+                if force && err.code==Base.UV_ENOENT
+                    return
+                end
+                # convert UVError to SystemError
+                rethrow(SystemError(err.prefix, err.code, path))
```

Comment:
On Thu, Dec 3, 2015 at 11:44 PM, Tony Kelman <notifications@github.com>
wrote:

> In base/file.jl
> <https://github.com/JuliaLang/julia/pull/14248#discussion_r46648880>:
>
> >      if islink(path) || !isdir(path)
> > -        @windows_only if (filemode(path) & 0o222) == 0; chmod(path, 0o777); end # is writable on windows actually means "is deletable"
> > -        unlink(path)
> > +        try
> > +            @windows_only if (filemode(path) & 0o222) == 0; chmod(path, 0o777); end # is writable on windows actually means "is deletable"
> > +            unlink(path)
> > +        catch err
> > +            if isa(err, UVError)
> > +                if force && err.code==Base.UV_ENOENT
> > +                    return
> > +                end
> > +                # convert UVError to SystemError
> > +                rethrow(SystemError(err.prefix, err.code, path))
>
> most are negatives of the errno.h versions, but only on unix:
> https://github.com/JuliaLang/libuv/blame/julia-uv0.11.26/include/uv-errno.h
>
> my errno_h.jl on Windows looks quite different:
> https://gist.github.com/b3e80b58153607dfb9a9
>
> kind of a shame to lose the extrainfo of the path in the error, but we can
> leave that to a later PR to figure out
>
I guess the easiest thing to do later on is to create a `SystemError` from
scratch, which would re-derive the error code from the current `errno`.

-erik

-- 
Erik Schnetter <schnetter@gmail.com>
http://www.perimeterinstitute.ca/personal/eschnetter/